File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int>hash(256,-1);
        int left=0;
        int maxLen=0;
        for(int right=0;right<s.size();right++){
            if(hash[s[right]]!=-1){
                left=max(left,hash[s[right]]+1);
            }
            hash[s[right]]=right;
            maxLen=max(maxLen,right-left+1);
        }
        return maxLen;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character,Integer>mpp=new HashMap<>();
        int left=0;
        int maxLen=0;
        for(int right=0;right<s.length();right++){
            mpp.put(s.charAt(right),mpp.getOrDefault(s.charAt(right),0)+1);
            while(mpp.get(s.charAt(right))>1){
                mpp.put(s.charAt(left),mpp.get(s.charAt(left))-1);
                if(mpp.get(s.charAt(left))==0){
                    mpp.remove(s.charAt(left));
                }
                left++;
            }
            maxLen=Math.max(maxLen,right-left+1);
        }
        return maxLen;
    }
}